sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkParserUtils.scala

```diff
@@ -50,22 +43,64 @@ trait SparkParserUtils {
       }
     }
 
-    if (b.startsWith("r") || b.startsWith("R")) {
+    def allCharsAreHex(s: String, start: Int, length: Int): Boolean = {
+      val end = start + length - 1
+      var i = start
+      while (i < end) {
+        val c = s.charAt(i)
+        val cIsHex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
```

Comment:
This is the same logic as `lang3.CharUtils#isAsciiAlphanumeric()`, but currently is ok

Reply:
It is not exactly equivalent: `isAsciiAlphanumeric` matches more alphabetic characters beyond those use in hexadecimal. I believe that the existing code would not unescape for invalid inputs (e.g. `\uZZZZ`) but I don't see an explicit test case for that, so I'll add one in order to more fully specify the existing behavior.

Reply:
Added new test cases for this in https://github.com/apache/spark/pull/47062/commits/972cf70d4306108a8357c0305fd8cec7d26ad233 and confirmed that they behaved the same even before this PR's changes.